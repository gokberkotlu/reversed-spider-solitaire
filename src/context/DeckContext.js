import React, { useState, createContext, useEffect } from "react";
import { cardSeries } from '../utils/CardSeries';

export const DeckContext = createContext();

export const DeckProvider = props => {

    // States
    const [gameCards, setGameCards] = useState([[],[],[],[],[],[],[],[],[],[]]); // Represents cards on table
    const [remainingCards, setRemainingCards] = useState([]); // Cards outside the game
    const [numberOfCompleted, setNumberOfCompleted] = useState(Array(0)); // Shows number of completed series
    const [newCardPacks, setNewCardPacks] = useState(Array(5)); // Shows number of new card packs
    const [isGameStarted, setIsGameStarted] = useState(false); // Set state when any card action happened
    const [statesLog, setStatesLog] = useState([]); // Log States to Undo the Game

    // States: Score Datas
    const [score, setScore] = useState(500); // Game score which starts 500
    const [highScore, setHighScore] = useState(+localStorage.getItem('highScore')); // Recorded highscore to store in localStorage

    const dealTheCards = (cards, numberOfCards) => {
        
        // Save Changes When the New Card Pack Added
        logGameChanges();

        if(cards.length >= 10) {
            let gameSet = [...gameCards];
            for(let i = 0; i < numberOfCards; i++) {
                gameSet[i%10].push(cards[i]);
            }
            setGameCards(gameSet);
            setRemainingCards(cards.splice(numberOfCards));

            return gameSet;
        }
    }

    // arrangeReversedCards arranges the reversed card at the start
    const arrangeReversedCards = (gameSet) => {
        gameSet.forEach(cards => {
            for(let i = 0; i < cards.length - 1; i++) {
                cards[i].isCardReverse = true;
            }
        })
    }

    // suffleTheCards shuffles eight decks of cards
    const suffleTheCards = () => {
        let deskNumber = 8;
        let newArray = [];
        for(let i = 0; i < deskNumber; i++) {
            newArray = newArray.concat(
                [...cardSeries.map((card) => {
                    return {
                        "cardNum": card,
                        "cardDeck": (i + 1),
                        "isCardReverse": false
                    }
                })]);
        }
        var currentIndex = newArray.length,
            randomIndex;
        // While there remain elements to shuffle...
        while (0 !== currentIndex) {
            // Pick a remaining element...
            randomIndex = Math.floor(Math.random() * currentIndex);
            currentIndex--;
    
            // And swap it with the current element.
            [newArray[currentIndex], newArray[randomIndex]] = [
                newArray[randomIndex], newArray[currentIndex]
            ];
        }
        return newArray;
    }

    // logGameChanges takes snapshots before all card actions and stores them inside statesLog
    const logGameChanges = () => {
        let addToLog = {
            "gameCardsBefore": JSON.parse(JSON.stringify(gameCards)),
            "remainingCardsBefore": JSON.parse(JSON.stringify(remainingCards)),
            "numberOfCompletedBefore": JSON.parse(JSON.stringify(numberOfCompleted)),
            "newCardPacksBefore": JSON.parse(JSON.stringify(newCardPacks))
        }
        setStatesLog(prevState => [...prevState, addToLog]);
    }

    // undoTheGame sets game states with last logged snapshot
    const undoTheGame = () => {
        let statesLogValue = [...statesLog];
        let statesLogValueLength = statesLogValue.length
        if(statesLogValueLength > 1) {
            let beforeStates = statesLogValue[statesLogValueLength - 1]; // Go back in the game by having "before value of last screen"
            setGameCards(beforeStates.gameCardsBefore);
            setRemainingCards(beforeStates.remainingCardsBefore);
            setNumberOfCompleted(beforeStates.numberOfCompletedBefore);
            setNewCardPacks(beforeStates.newCardPacksBefore);

            // Delete current states from log when went back in the game
            setStatesLog(statesLogValue.slice(0, statesLogValueLength - 1));

            // Decrease 1 point from score for each undo click
            setScore(prevState => prevState - 1);

            // Decrease 100 points from score if completed desk is removed by clicking undo button
            if(beforeStates.numberOfCompletedBefore.length < numberOfCompleted.length) {
                let scoreDiff = numberOfCompleted.length - beforeStates.numberOfCompletedBefore.length;
                setScore(prevState => prevState - 100 * scoreDiff);
            }
        }
    }

    useEffect(() => {
        // Shuffle cards, arrange reversed cards and deal them on the game table at start
        var allCards = suffleTheCards();
        var gameSet = dealTheCards(allCards, 54);
        arrangeReversedCards(gameSet);
    }, [])

    return (
        <DeckContext.Provider
        value={{
                "gameCards": [gameCards, setGameCards],
                "remainingCards": [remainingCards, setRemainingCards],
                "numberOfCompleted": [numberOfCompleted, setNumberOfCompleted],
                "newCardPacks": [newCardPacks, setNewCardPacks],
                "isGameStarted": [isGameStarted, setIsGameStarted],
                "suffleTheCards": suffleTheCards,
                "dealTheCards": dealTheCards,
                "arrangeReversedCards": arrangeReversedCards,
                "statesLog": [statesLog, setStatesLog],
                "logGameChanges": logGameChanges,
                "undoTheGame": undoTheGame,
                "score": [score, setScore],
                "highScore": [highScore, setHighScore]
            }}>
            {props.children}
        </DeckContext.Provider>
    );
}