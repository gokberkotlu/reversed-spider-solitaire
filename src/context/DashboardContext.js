import React, { useState, createContext, useContext, useEffect } from "react";
import { DeckContext } from "./DeckContext";

export const DashboardContext = createContext();

export const DashboardProvider = props => {

    // States
    const [counter, setCounter] = useState("00:00:00"); // Represents timer
    const [isGameFinished, setIsGameFinished] = useState(false); // According to this state, popup is shown or not
    const [resetBit, setResetBit] = useState(false); // It triggers the useEffect to reset the game
    const [decreaseInEveryTenSeconds, setDecreaseInEveryTenSeconds] = useState(0); // Decrease 2 points from score for each 10 sec

    // DeckContext
    const deckContextValue = useContext(DeckContext);

    const [, setGameCards] = deckContextValue.gameCards;
    const [, setRemainingCards] = deckContextValue.remainingCards;
    const [, setNumberOfCompleted] = deckContextValue.numberOfCompleted;
    const [, setNewCardPacks] = deckContextValue.newCardPacks;
    const [, setIsGameStarted] = deckContextValue.isGameStarted;
    const [, setStatesLog] = deckContextValue.statesLog;
    const suffleTheCards = deckContextValue.suffleTheCards;
    const dealTheCards = deckContextValue.dealTheCards;
    const arrangeReversedCards = deckContextValue.arrangeReversedCards;

    const [, setScore] = deckContextValue.score;

    const resetGame = () => {
        // Resetting
        setGameCards([[],[],[],[],[],[],[],[],[],[]]);
        setRemainingCards([]);
        setNewCardPacks(Array(5));
        setNumberOfCompleted(Array(0));
        setScore(500);
        setCounter("00:00:00");
        setIsGameFinished(false);
        setResetBit(true);
        setIsGameStarted(false);
        setStatesLog([]);
        setDecreaseInEveryTenSeconds(0);
    }

    useEffect(() => {// Deal cards again from scratch
        if(resetBit) {
            var allCards = suffleTheCards();
            var gameSet = dealTheCards(allCards, 54); // setGameCards and setRemainingCards included
            arrangeReversedCards(gameSet);
            setResetBit(false);
        }
    }, [resetBit]) // resetBit changes trigger the control of gameSet and remainingCards

    return (
        <DashboardContext.Provider
        value={{"counter": [counter, setCounter],
        "isGameFinished": [isGameFinished, setIsGameFinished],
        "resetGame": resetGame,
        "decreaseInEveryTenSeconds": [decreaseInEveryTenSeconds ,setDecreaseInEveryTenSeconds]}}>
            {props.children}
        </DashboardContext.Provider>
    );
}