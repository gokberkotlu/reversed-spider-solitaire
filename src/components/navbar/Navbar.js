import React, { useContext, useEffect } from 'react';
import { DeckContext } from "../../context/DeckContext";
import { DashboardContext } from "../../context/DashboardContext";
import "./navbar.css";
import timerIcon from "../../icons/Timer.svg";
import scoreIcon from "../../icons/Score.svg";
import newGameIcon from "../../icons/NewGame.svg";
import undoIcon from "../../icons/Undo.svg";

const Navbar = () => {

    // DeckContext
    const deckContextValue = useContext(DeckContext);

    const [isGameStarted,] = deckContextValue.isGameStarted;
    const undoTheGame = deckContextValue.undoTheGame;
    const [score, setScore] = deckContextValue.score;
    const [highScore,] = deckContextValue.highScore;

    // DashboardContext
    const dashboardContextValue = useContext(DashboardContext);

    const [counter, setCounter] = dashboardContextValue.counter;
    const [decreaseInEveryTenSeconds, setDecreaseInEveryTenSeconds] = dashboardContextValue.decreaseInEveryTenSeconds;
    const resetGame = dashboardContextValue.resetGame;

    useEffect(() => {
        // Set timer with current rules
        if(isGameStarted) {
            const timer = setInterval(() => {
                let timeData = counter.split(":")
                let hour = +timeData[0];
                let minute = +timeData[1];
                let second = +timeData[2] + 1;

                // Show seconds as two digits
                if(second < 10) {
                    second = '0' + second;
                } else if(second > 59) {
                    second = '00';
                    minute = minute + 1;
                }

                // Show minutes as two digits
                if(minute < 10) {
                    minute = '0' + minute;
                } else if(minute > 59) {
                    minute = '00';
                    hour = hour + 1;
                }

                // Show hours as two digits
                if(hour < 10) {
                    hour = '0' + hour;
                }

                setCounter(`${hour}:${minute}:${second}`); // set counter with new time data

                setDecreaseInEveryTenSeconds(prevState => prevState + 1); // Increase state until it is ten

            }, 1000);
            return () => clearInterval(timer);
        }
    }, [counter, isGameStarted]);

    useEffect(() => {
        // Decrease score by two in every ten seconds
        if(decreaseInEveryTenSeconds >= 10) {
            setDecreaseInEveryTenSeconds(0);
            setScore(prevState => prevState - 2);
        }
    }, [decreaseInEveryTenSeconds])

    return (
        <nav>
            <div className="navbar-container">
                <div className="score-timer-container">
                    <div className="timer-container">
                        <img src={timerIcon} className="navbar-icon" width={20} height={20} alt="Timer" />
                        <span>{counter}</span>
                    </div>
                    <div className="score-container">
                        <img src={scoreIcon} className="navbar-icon" width={20} height={20} alt="Score" />
                        <span>{ score }</span>
                    </div>
                </div>
                <span>High Score: { (highScore === null) ? 0 : highScore }</span>
                <div className="undo-new-container">
                    <div className="undo-container">
                        <div className="navbar-btn" onClick={() => {undoTheGame()}}>
                            <img src={undoIcon} className="navbar-icon" width={20} height={20} alt="Undo" />
                            <span>Undo</span>
                        </div>
                    </div>
                    <div className="new-container">
                        <div className="navbar-btn" onClick={() => {resetGame()}}>
                            <img src={newGameIcon} className="navbar-icon" width={20} height={20} alt="New Game" />
                            <span>New</span>
                        </div>
                    </div>
                </div>
            </div>
        </nav>
    );
}
 
export default Navbar;