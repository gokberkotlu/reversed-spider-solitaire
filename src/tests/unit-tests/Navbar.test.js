import { screen, render } from '@testing-library/react';
import renderer from "react-test-renderer";
import Navbar from '../../components/navbar/Navbar';
import { DeckProvider } from '../../context/DeckContext';
import { DashboardProvider } from "../../context/DashboardContext"

describe('Navbar component', () => {
    it('should render Navbar component', () => {
        const tree = renderer.create(
            <DeckProvider value={null}>
                <DashboardProvider value={null}>
                    <Navbar />
                </DashboardProvider>
            </DeckProvider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('should have timer at beginning state(00:00:00)', () => {
        render(
            <DeckProvider value={null}>
                <DashboardProvider value={null}>
                    <Navbar />
                </DashboardProvider>
            </DeckProvider>
        );
        expect(screen.getByText(`00:00:00`)).toBeInTheDocument();
    })

    it('should have score at beginning state(500) and score should wrapped by score-container class', () => {
        render(
            <DeckProvider value={null}>
                <DashboardProvider value={null}>
                    <Navbar />
                </DashboardProvider>
            </DeckProvider>
        );
        expect(screen.getByText(`500`).parentElement.className).toBe("score-container");
    })

    it('should have 0 points highscore at beginning', () => {
        render(
            <DeckProvider value={null}>
                <DashboardProvider value={null}>
                    <Navbar />
                </DashboardProvider>
            </DeckProvider>
        );
        expect(screen.getByText(`High Score: 0`)).toBeTruthy();
    })

    it('should have undo button in the document', () => {
        render(
            <DeckProvider value={null}>
                <DashboardProvider value={null}>
                    <Navbar />
                </DashboardProvider>
            </DeckProvider>
        );
        const undoButton = screen.getByText(/Undo/i).parentElement;
        expect(undoButton).toBeInTheDocument();
    })

    it('should have new button in the document', () => {
        render(
            <DeckProvider value={null}>
                <DashboardProvider value={null}>
                    <Navbar />
                </DashboardProvider>
            </DeckProvider>
        );
        const newButton = screen.getByText(/New/i).parentElement;
        expect(newButton).toBeInTheDocument();
    })
})