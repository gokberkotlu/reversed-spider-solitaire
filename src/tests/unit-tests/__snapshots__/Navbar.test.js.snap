// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component should render Navbar component 1`] = `
<nav>
  <div
    className="navbar-container"
  >
    <div
      className="score-timer-container"
    >
      <div
        className="timer-container"
      >
        <img
          alt="Timer"
          className="navbar-icon"
          height={20}
          src="Timer.svg"
          width={20}
        />
        <span>
          00:00:00
        </span>
      </div>
      <div
        className="score-container"
      >
        <img
          alt="Score"
          className="navbar-icon"
          height={20}
          src="Score.svg"
          width={20}
        />
        <span>
          500
        </span>
      </div>
    </div>
    <span>
      High Score: 
      0
    </span>
    <div
      className="undo-new-container"
    >
      <div
        className="undo-container"
      >
        <div
          className="navbar-btn"
          onClick={[Function]}
        >
          <img
            alt="Undo"
            className="navbar-icon"
            height={20}
            src="Undo.svg"
            width={20}
          />
          <span>
            Undo
          </span>
        </div>
      </div>
      <div
        className="new-container"
      >
        <div
          className="navbar-btn"
          onClick={[Function]}
        >
          <img
            alt="New Game"
            className="navbar-icon"
            height={20}
            src="NewGame.svg"
            width={20}
          />
          <span>
            New
          </span>
        </div>
      </div>
    </div>
  </div>
</nav>
`;
